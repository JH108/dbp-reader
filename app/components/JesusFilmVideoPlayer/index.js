/**
 *
 * JesusFilmVideoPlayer
 *
 */

import React from 'react';
import PropTypes from 'prop-types';
import Router from 'next/router';
import SvgWrapper from '../SvgWrapper';
import VideoControls from '../VideoControls';
import VideoProgressBar from '../VideoProgressBar';
import VideoOverlay from '../VideoOverlay';

class JesusFilmVideoPlayer extends React.PureComponent {
  // eslint-disable-line react/prefer-stateless-function
  state = {
    paused: true,
    volume: 1,
    currentTime: 0,
    bufferLength: 0,
    hlsSupported: true,
  };

  componentDidMount() {
    if (this.videoRef) {
      this.getHls();
      this.videoRef.addEventListener(
        'webkitendfullscreen',
        this.webkitendfullscreen,
        false,
      );
      Router.router.events.on('routeChangeStart', this.handleRouteChange);
    }
  }

  componentWillUnmount() {
    if (this.hls && this.hls.media) {
      this.hls.media.removeEventListener(
        'timeupdate',
        this.timeUpdateEventListener,
      );
      this.hls.media.removeEventListener('seeking', this.seekingEventListener);
      this.hls.media.removeEventListener('seeked', this.seekedEventListener);
      this.hls.detachMedia();
      this.hls.stopLoad();
      this.hls.destroy();
    }
    if (this.videoRef) {
      this.videoRef.removeEventListener(
        'timeupdate',
        this.timeUpdateEventListener,
      );
      this.videoRef.removeEventListener('seeking', this.seekingEventListener);
      this.videoRef.removeEventListener('seeked', this.seekedEventListener);
      this.videoRef.removeEventListener(
        'webkitendfullscreen',
        this.webkitendfullscreen,
      );
      this.videoRef.removeEventListener('loadedmetadata', this.loadedMetadata);
    }

    Router.router.events.off('routeChangeStart', this.handleRouteChange);
  }

  webkitendfullscreen = () => {
    this.pauseVideo();
  };

  getHls = async () => {
    const hls = await import('hls.js');

    this.Hls = hls.default;
    this.isSupported = hls.isSupported;

    if (this.videoRef) {
      this.checkHlsSupport();
      this.initVideoStream();
    }
  };

  get playButton() {
    const { paused } = this.state;

    return (
      <SvgWrapper
        onClick={this.playVideo}
        className={paused ? 'play-video show-play' : 'play-video hide-play'}
        fill={'#fff'}
        svgid={'play_video'}
        viewBox={'0 0 90 40'}
      />
    );
  }

  setVideoRef = (el) => {
    this.videoRef = el;
  };

  setBuffer = () => {
    // Can accept current time as the first parameter
    if (this.hls && this.hls.media) {
      const buf = this.hls.media.buffered;
      if (buf && buf.length) {
        this.setState({ bufferLength: buf.end(buf.length - 1) });
      }
    }
  };

  setCurrentTime = (time) => {
    if (this.hls.media) {
      this.hls.media.currentTime = time;
      this.setState({ currentTime: time });
    } else {
      this.videoRef.currentTime = time;
      this.setState({ currentTime: time });
    }
  };

  handleRouteChange = () => {
    if (this.hls) {
      this.hls.destroy();
    }
    if (this.videoRef) {
      this.pauseVideo();
    }
  };

  handleVideoClick = () => {
    const { paused } = this.state;

    if (paused) {
      this.playVideo();
    } else {
      this.pauseVideo();
    }
  };

  checkHlsSupport = () => {
    if (typeof this.isSupported === 'function') {
      this.setState({
        hlsSupported: this.isSupported(),
      });
    } else {
      this.setState({
        hlsSupported: false,
      });
    }
  };

  initHls = () => {
    // Destroying the old hls stream so that there aren't artifacts leftover in the new stream
    if (this.hls) {
      this.hls.destroy();
    }
    this.hls = new this.Hls();
    this.hls.on(this.Hls.Events.ERROR, (event, data) => {
      if (data.fatal) {
        switch (data.type) {
          case this.Hls.ErrorTypes.NETWORK_ERROR:
            this.hls.startLoad();
            break;
          case this.Hls.ErrorTypes.MEDIA_ERROR:
            this.hls.recoverMediaError();
            break;
          default:
            this.hls.destroy();
            break;
        }
      }
    });
  };

  loadedMetadata = () => {
    this.videoRef.play();
    this.setState({ paused: false });
  };

  initVideoStream = () => {
    const { hlsSupported } = this.state;
    const { hlsStream } = this.props;
    if (!hlsSupported) {
      if (this.videoRef.canPlayType('application/vnd.apple.mpegurl')) {
        this.videoRef.src = `${hlsStream}?key=${
          process.env.DBP_API_KEY
        }&v=4&asset_id=dbp-vid`;
        this.videoRef.addEventListener(
          'timeupdate',
          this.timeUpdateEventListener,
        );
        this.videoRef.addEventListener('seeking', this.seekingEventListener);
        this.videoRef.addEventListener('seeked', this.seekedEventListener);
      }
    } else {
      // Create the hls stream first
      this.initHls();
      try {
        // Check for the video element
        if (this.videoRef) {
          // Make sure that there is a valid source
          if (hlsStream) {
            this.hls.attachMedia(this.videoRef);
            this.hls.loadSource(
              `${hlsStream}?key=${
                process.env.DBP_API_KEY
              }&v=4&asset_id=dbp-vid`,
            );
            this.hls.media.addEventListener(
              'timeupdate',
              this.timeUpdateEventListener,
            );
            this.hls.media.addEventListener(
              'seeking',
              this.seekingEventListener,
            );
            this.hls.media.addEventListener('seeked', this.seekedEventListener);
            this.hls.on(this.Hls.Events.MANIFEST_PARSED, () => {
              this.hls.media.play();
              this.setState({ paused: false });
            });
            this.hls.on(this.Hls.Events.BUFFER_APPENDING, () => {
              this.setBuffer();
            });
          }
        }
      } catch (err) {
        if (process.env.NODE_ENV === 'development') {
          console.error('initVideoStream', err); // eslint-disable-line no-console
        }
      }
    }
  };

  timeUpdateEventListener = (e) => {
    this.setState({
      currentTime: e.target.currentTime,
    });
  };

  seekingEventListener = (e) => {
    this.setState({
      currentTime: e.target.currentTime,
    });
  };

  seekedEventListener = (e) => {
    this.setState({
      currentTime: e.target.currentTime,
    });
  };

  playVideo = ({ thumbnailClick }) => {
    const { hlsSupported } = this.state;
    const { hlsStream } = this.props;
    if (!hlsSupported) {
      if (
        this.videoRef.src ===
        `${hlsStream}?key=${process.env.DBP_API_KEY}&v=4&asset_id=dbp-vid`
      ) {
        this.videoRef.play();
        this.setState({ paused: false });
        // if the sources didn't match then this is a new video and the hls stream needs to be updated
      } else {
        // Init a new hls stream
        this.initVideoStream({ thumbnailClick });
      }
    } else {
      try {
        // if the current video has a source (initial load may be an empty object)
        if (hlsStream) {
          // if there is already an hls stream and that streams url is equal to this videos source then play the video
          if (
            this.hls.media &&
            this.hls.url ===
              `${hlsStream}?key=${process.env.DBP_API_KEY}&v=4&asset_id=dbp-vid`
          ) {
            this.hls.media.play();
            this.setState({ paused: false });
            // if the sources didn't match then this is a new video and the hls stream needs to be updated
          } else {
            // Stop the current player from loading anymore video
            this.hls.stopLoad();
            // Remove the old hls stream
            this.hls.destroy();
            // Init a new hls stream
            this.initVideoStream({ thumbnailClick });
          }
        }
      } catch (err) {
        if (process.env.NODE_ENV === 'development') {
          console.error('caught in playVideo', err); // eslint-disable-line no-console
        }
      }
    }
  };

  pauseVideo = () => {
    this.videoRef.pause();
    this.setState({ paused: true });
  };

  toggleFullScreen = () => {
    const isFullScreen = !!(
      document.fullScreen ||
      document.webkitIsFullScreen ||
      document.mozFullScreen ||
      document.msFullscreenElement ||
      document.fullscreenElement
    );

    if (isFullScreen) {
      if (document.exitFullscreen) {
        document.exitFullscreen();
      } else if (document.mozCancelFullScreen) {
        document.mozCancelFullScreen();
      } else if (document.webkitCancelFullScreen) {
        document.webkitCancelFullScreen();
      } else if (document.msExitFullscreen) {
        document.msExitFullscreen();
      }
    } else if (this.videoRef) {
      if (this.videoRef.requestFullscreen) {
        this.videoRef.requestFullscreen();
      } else if (this.videoRef.mozRequestFullScreen) {
        this.videoRef.mozRequestFullScreen();
      } else if (this.videoRef.webkitRequestFullScreen) {
        this.videoRef.webkitRequestFullScreen();
      } else if (this.videoRef.msRequestFullscreen) {
        this.videoRef.msRequestFullscreen();
      }
    }
  };

  updateVolume = (volume) => {
    this.videoRef.volume = volume;
    this.setState({ volume });
  };

  render() {
    const { volume, paused, currentTime, bufferLength } = this.state;
    const { hasVideo, hlsStream, duration } = this.props;
    // console.log('hlsStream', hlsStream, 'duration', duration);
    // Don't render anything if there is no video for the chapter
    if (!hasVideo || !hlsStream) {
      return (
        <div
          key={'video-player-container'}
          className={'video-player-container jesus-film-override'}
        >
          <h2>
            Something went wrong. Please refresh the page or go back to try
            again!
          </h2>
        </div>
      );
    }
    /* eslint-disable jsx-a11y/media-has-caption */
    return (
      <div
        key={'video-player-container'}
        className={'video-player-container jesus-film-override'}
      >
        <div className={'video-player'}>
          <VideoOverlay
            paused={paused}
            playFunction={this.playVideo}
            pauseFunction={this.pauseVideo}
            isJesusFilm
          />
          <video ref={this.setVideoRef} onClick={this.handleVideoClick} />
          <VideoProgressBar
            paused={paused}
            currentTime={currentTime}
            duration={duration || 300}
            setCurrentTime={this.setCurrentTime}
            bufferLength={bufferLength}
          />
          <VideoControls
            paused={paused}
            pauseVideo={this.pauseVideo}
            toggleFullScreen={this.toggleFullScreen}
            updateVolume={this.updateVolume}
            volume={volume}
          />
        </div>
      </div>
    );
    /* eslint-enable jsx-a11y/media-has-caption */
  }
}

JesusFilmVideoPlayer.propTypes = {
  hlsStream: PropTypes.string.isRequired,
  duration: PropTypes.number.isRequired,
  hasVideo: PropTypes.bool.isRequired,
};

export default JesusFilmVideoPlayer;
