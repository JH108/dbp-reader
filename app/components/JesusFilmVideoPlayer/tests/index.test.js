import React from 'react';
import renderer from 'react-test-renderer';

import JesusFilmVideoPlayer from '..';

const props = {
  hlsStream:
    'https://api-dev.dbp4.org/arclight/jesus-film?key=e8a946a0-d9e2-11e7-bfa7-b1fb2d7f5824&v=4&arclight_id=23156',
  duration: 5789,
  hasVideo: true,
};

describe('<JesusFilmVideoPlayer /> component tests', () => {
  it('Should match snapshot with default props', () => {
    const tree = renderer.create(<JesusFilmVideoPlayer {...props} />).toJSON();

    expect(tree).toMatchSnapshot();
  });
  it('Should match snapshot with no hls stream', () => {
    const tree = renderer
      .create(<JesusFilmVideoPlayer {...props} hlsStream={''} />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
