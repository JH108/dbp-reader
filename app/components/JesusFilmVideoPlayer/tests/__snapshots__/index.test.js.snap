// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<JesusFilmVideoPlayer /> component tests Should match snapshot with default props 1`] = `
<div
  className="video-player-container jesus-film-override"
>
  <div
    className="video-player"
  >
    <div
      className="play-video-container show-control-icon"
      onClick={[Function]}
    >
      <div
        className="control-button-container"
        onClick={[Function]}
      >
        <span
          className="play-video-title"
        >
          Jesus Film
        </span>
        <svg
          className="play-video"
          fill="#fff"
        >
          <use
            xlinkHref="/static/svglist.svg#play_video"
            xmlnsXlink="http://www.w3.org/1999/xlink"
          />
        </svg>
      </div>
    </div>
    <video
      onClick={[Function]}
    />
    <div
      className="progress-slider hide-progress-slider"
      data-value-cur="00:00"
      data-value-dur="01:36:29"
    >
      <div
        className="progress-bar-buffer"
        style={
          Object {
            "width": "0%",
          }
        }
      />
      <div
        className="rc-slider"
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        onMouseUp={[Function]}
        onTouchStart={[Function]}
      >
        <div
          className="rc-slider-rail"
          style={
            Object {
              "backgroundColor": "rgba(199, 199, 204, 0.95)",
              "cursor": "pointer",
              "height": "2px",
            }
          }
        />
        <div
          className="rc-slider-track"
          style={
            Object {
              "backgroundColor": "rgba(98, 177, 130, 1)",
              "cursor": "pointer",
              "height": "2px",
              "left": "0%",
              "width": "0%",
            }
          }
        />
        <div
          className="rc-slider-step"
        />
        <div
          aria-disabled={false}
          aria-valuemax={100}
          aria-valuemin={0}
          aria-valuenow={0}
          className="rc-slider-handle"
          onBlur={[Function]}
          onKeyDown={[Function]}
          role="slider"
          style={
            Object {
              "backgroundColor": "rgba(98, 177, 130, 1)",
              "border": "2px solid rgba(98, 177, 130, 1)",
              "height": "10px",
              "left": "0%",
              "top": "6px",
              "width": "10px",
            }
          }
          tabIndex={0}
        />
        <div
          className="rc-slider-mark"
        />
      </div>
    </div>
    <div
      className="controls hide-controls"
    >
      <div
        className="left-controls"
      >
        <div
          className="video-volume-container"
          onClick={[Function]}
          onTouchEnd={[Function]}
        >
          <div
            className="video-volume-slider-container"
          >
            <div
              className="video-slider-container"
            >
              <div
                className="rc-slider rc-slider-vertical slider"
                onBlur={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onMouseDown={[Function]}
                onMouseUp={[Function]}
                onTouchStart={[Function]}
              >
                <div
                  className="rc-slider-rail"
                  style={
                    Object {
                      "backgroundColor": "#000",
                      "width": "2px",
                    }
                  }
                />
                <div
                  className="rc-slider-track"
                  style={
                    Object {
                      "backgroundColor": "rgba(98, 177, 130, 1)",
                      "bottom": "0%",
                      "height": "100%",
                      "width": "2px",
                    }
                  }
                />
                <div
                  className="rc-slider-step"
                />
                <div
                  aria-disabled={false}
                  aria-valuemax={100}
                  aria-valuemin={0}
                  aria-valuenow={100}
                  className="rc-slider-handle"
                  onBlur={[Function]}
                  onKeyDown={[Function]}
                  role="slider"
                  style={
                    Object {
                      "backgroundColor": "rgba(98, 177, 130, 1)",
                      "borderColor": "rgba(98, 177, 130, 1)",
                      "bottom": "100%",
                      "height": "12.5px",
                      "left": "5px",
                      "width": "12.5px",
                    }
                  }
                  tabIndex={0}
                />
                <div
                  className="rc-slider-mark"
                />
              </div>
            </div>
          </div>
          <svg
            className="icon"
            fill="#fff"
          >
            <use
              xlinkHref="/static/svglist.svg#volume_max"
              xmlnsXlink="http://www.w3.org/1999/xlink"
            />
          </svg>
        </div>
        <div
          onClick={[Function]}
        >
          <svg
            fill="#fff"
          >
            <use
              xlinkHref="/static/svglist.svg#pause"
              xmlnsXlink="http://www.w3.org/1999/xlink"
            />
          </svg>
        </div>
      </div>
      <div
        className="right-controls"
      >
        <div>
          <svg
            className="video-elipsis"
            fill="#fff"
          >
            <use
              xlinkHref="/static/svglist.svg#elipsis"
              xmlnsXlink="http://www.w3.org/1999/xlink"
            />
          </svg>
        </div>
        <div
          onClick={[Function]}
        >
          <svg
            className="video-fullscreen"
            fill="#fff"
          >
            <use
              xlinkHref="/static/svglist.svg#fullscreen"
              xmlnsXlink="http://www.w3.org/1999/xlink"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<JesusFilmVideoPlayer /> component tests Should match snapshot with no hls stream 1`] = `
<div
  className="video-player-container jesus-film-override"
>
  <h2>
    Something went wrong. Please refresh the page or go back to try again!
  </h2>
</div>
`;
